### PR TITLE
[Enhancement] support array_agg MV rewrite

### PR DESCRIPTION
**Pull request by murphyatwork**
## What I'm doing:
- Support rewrite `count(distinct)` to `array_agg_distinct` function
- Direct rewrite: `count(distinct x) => array_length(xs)` 
- Rollup rewrite: `count(distinct x) => array_length(array_unique_agg(xs))` 

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorUtil.java [hunk 1]
```diff
@@ -64,6 +65,10 @@ public static Function findArithmeticFunction(Type[] argsType, String fnName) {
         return Expr.getBuiltinFunction(fnName, argsType, IS_IDENTICAL);
     }
 
+    public static Function findRolluFunction(Type[] argsType, String fnName) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorUtil.java:68]: typo

Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorUtil.java]: fixed

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java [hunk 2]
```diff
@@ -646,8 +649,43 @@ private Map<ColumnRefOperator, CallOperator> rewriteAggregates(Map<ColumnRefOper
...
+                                .create(realAggregate, realAggregate.getType(), realAggregate.isNullable());
+                CallOperator copyProject = (CallOperator) newAggregate.clone();
+                copyProject.setChild(foundIndex, innerAgg);
+
+                ColumnRefOperator outerProject = context.getQueryRefFactory()
+                                .create(copyProject, copyProject.getType(), copyProject.isNullable());
+                newProjection.put(outerProject, copyProject);
+                newAggregations.put(innerAgg, realAggregate);
+                // replace original projection
+                aggregateMapping.put(entry.getKey(), copyProject);
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:684]: Good Catch.

**Review by LiShuMing — APPROVED** [review]
LGTM

**Review by ABingHuang — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Pull request merged**